---
title: "P/BIO381 Tutorials: Intro to connecting to unix servers and navigating the bash command-line"
date: 'January 22, 2020'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

*What is the command-line?*

The bash command line, also known as a "terminal" or "shell", is a way of interacting with your local computer or a remote server by means of typing commands or scripts, without using a graphical user interface (GUI).



*Why do I want to be doing this?*

At first, the command line can seem a little intimidating. But after you get used to typing instead of pointing and clicking to issue your commands, you'll realize how powerful it is. For example, it's quite easy to copy, move, edit, and search within thousands of files in multiple directories with some simple command line code. It would take forever to do this by dragging/dropping with a mouse. The command line also allows you to work with very large data files without uncompressing them fully, or loading the entire file's contents into memory…something that standard GUI type applications aren't good at.



*So, let's get started…*

- The first step is to open a terminal *shell* on your local computer. For windows users, this would be "PuTTy". For MacOS users, this is called "Terminal".

- We'll connect to our remote server running Unix using the secure shell (ssh) protocol. Our server's name is *pbio381* and we can connect to it using our UVM netid username and password (as long as we're on-campus)

```
ip0af52fbf:papers srkeller$ ssh srkeller@pbio381.uvm.edu
srkeller@pbio381.uvm.edu's password: 
Last login: Tue Jan 21 14:39:33 2020 from 132.198.159.152
```


- The log-in screen tells us some basic info on when we last logged in, and then gives us our current location in the filesystem (~) followed by the $ prompt, that tells us the computer is ready for our next command. 

  - NOTE: The tilda (~) is short-hand for your home directory in UNIX. This is your own personal little corner of the computer's hard drive space, and is the location that you should use to create folders and input/output/results files that are specific to your own work. No one has access to any files stored in your home directory but you.

- To see the full path to your current directory, use the `pwd` command:

```
  [srkeller@pbio381 ~]$ pwd
  /users/s/r/srkeller
  [srkeller@pbio381 ~]$ 
```

- The path shows the full directory address up from the "root" of the file structure, which is the most basal level (appealing to all you phylogeneticists here…). The root is symbolized as "/" and each subdirectory is separated by an additional "/". So, the full path to my working directory on the server is `/users/s/r/srkeller/`

- We're going to use the Github repos we just created to help us stay organized with files moving back and forth between the server and our local machines.  The server is able to run git at the command line, which takes a little getting used to.  The basic commands all start with calling the program `git` followed by the type of command you want to do.  First, we're going to make a copy (i.e., clone) of our repo that's on Github (in the cloud) and install it into our home directory on the server.  You'll only have to do this once!  

- Go to your Github account in your web browser and navigate to your repo for Ecological Genomics.  Click the green "Clone or download" button on the right, and then click the icon of the clipboard.  The web address to your repo will be copied temporarily into memory, which you'll then paste in the command below on your terminal shell.  Your address should look like mine but with your username and repo name, and ending in ".git""

```
git clone https://github.com/stephenrkeller/Ecological_Genomics.git
```

- We now have a cloned copy of our repo on the server in your home directory. Now, all your work between the server and your laptop can be synchronized and kept up to date!  

- OK, back to bash.  Let's navigate into your repo (what's its name?) using the `cd` command

- We can then use the `ll` command to list out the contents of any folders and files:

```
[srkeller@pbio381 Spring_2020]$ ll
total 12
drwxr-xr-x. 3 srkeller users   49 Jan 21 14:29 mydata
drwxr-xr-x. 3 srkeller users   47 Jan 21 14:29 myresults
drwxr-xr-x. 3 srkeller users 4096 Jan 21 12:51 myscripts
-rw-r--r--. 1 srkeller users 7929 Jan 21 12:28 SRKeller_PBIO381_2020_online_notebook.md
```

- You'll notice that I've got some extra folders in my output from previous work, whereas you may only see your notebook.

- NOTE: Each row shows a file or a folder diplaying (from right to left) its name, when it was last edited, size (in bytes), who it belongs to (user & group), and who has permissions to read (r) write (w) and exectue (x) it. More on permissions later...

- You should make a new folder (aka, a directory, or "dir") using the `mkdir` command. Name this folder `mydata`

```
[srkeller@pbio381 ~]$ mkdir mydata
```

- Use the `ll` command to confirm it's there, then make two more folders named `myscripts` and `myresults`. Then use the `ll` command to list the folders again 

- We'll use this directory structure to store our scripts, data, and results. 

- For the first part of the semester, we're going to analyze exome-capture Illumina data sequenced from [red spruce (*Picea rubens* Sarg.)](https://en.wikipedia.org/wiki/Picea_rubens) trees.  Red spruce is a long-lived coniferous tree that grows in cool moist habitats like high elevations in the mountains or along the coast of Atlantic Canada.  To generate these data, members of the Keller Lab collected samples from trees across the Appalachian Mountains, from Tennessee, USA to New Brunswick, Canada. 

![](https://github.com/stephenrkeller/Ecological_Genomics/blob/master/Spring_2020/mydata/misc/Figure2_Capblancq_etal_EvolApps.png?raw=true) 
![](http://www.uvm.edu/~kellrlab/uploads/2/2/9/6/22961126/editor/img-20170930-155547560.jpg?1533919793)

- If it looks like this was a lot of hard work, you've got it right!


- We've placed the raw Illumina sequence files and a file containing the collection metadata under a shared directory on the server: `/data/project_data/RS_ExomeSeq`   

- Use `cd` to navigate over to this location. Then `ll` to show its contents. You should see something like this:
    
```
[srkeller@pbio381 ~]$ cd /data/project_data/RS_ExomeSeq
[srkeller@pbio381 mydata]$ ll
drwxr-xr-x. 3 srkeller users 31 Jan 16 13:16 fastq
drwxr-xr-x. 2 srkeller users 42 Jan 21 14:06 metadata
[srkeller@pbio381 mydata]$ 
```
- The `fastq` dir contains the paired-end Illumina sequence files, while the `metadata` dir contains the collection info.  Use `cd` to navigate inside the `metadata` folder and `ll`; you should see a file called `RS_Exome_metadata.txt`.  

- We don't want to open and make changes to this file in the shared space, because we don't want to have our edits affect the rest of the group.  Best practices are to make a **copy** of this file over to our home directory and put it inside your `mydata` folder. Use the `cp` command, followed by the filename, and the path to your destination (remember the ~ signals your home directory, and each subdirectory is then separated by a /):

```
[srkeller@pbio381 metadata]$ cp RS_Exome_metadata.txt ~/<YourGithubRepoName>/mydata/
```
- Now `cd` back to `~/<YourGithubRepoName>/mydata/` and look inside. You should see your file...

- We can use the `head` command to peek at the first 10 lines of data; or more generally, `head -n # filename` will show the first # lines of data:

```
[srkeller@pbio381 metadata]$ head RS_Exome_metadata.txt
[srkeller@pbio381 metadata]$ head -n 20 RS_Exome_metadata.txt
``` 

- The column headers should be self-explanatory.  Note the column 'Region' -- this refers to the geographical and genetic cluster that each individual belongs to (see map above):  the "Core" (C), "Margin" (M), or "Edge" (E).  This entire dataset is part of a larger NSF-funded project on the population genomics of climate adaptation in red spruce.  

  - NOTE: The `tail` command provides similar functionality as `head`, but prints just the *last lines in the file*. 
  
-For our class work, we're going to be focusing on **just the Edge samples**.  Bash gives us a convenient way to search within our file and pull out just the rows of data that match a term of interest (in this case, we'll search for "E").  We do this using a "generalized regular expression, also known as the `grep` command. Any line matching our search string will be printed to screen.  The "option "-w" option tells `grep` to match the entire field has to be matched as-is.

```
[srkeller@pbio381 mydata]$ grep -w "E" RS_Exome_metadata.txt
AB      05      AB_05   TN      E       09002016        35.55297        83.49438        1812    57.7
AB      08      AB_08   TN      E       09002016        35.55212        83.49259        1785    44.3
AB      12      AB_12   TN      E       09002016        35.5389 83.49463        1750    36.8
[...]
```

- What if instead of printing it to screen, we want to save the output of our search to a new file? This is easy, just use the ">" symbol to redirect the results of any command to an output file with your choice of name.

```
[srkeller@pbio381 mydata]$ grep -w "E" RS_Exome_metadata.txt >Edge_only.txt
```

- `Grep` is a useful search tool and has many additional features for sorting and output of the results. These kinds of search algorithms are called "regular expressions", or "regexp", and are one of the most powerful tools for wokring with large text files. If you want to learn more about `grep` and its regexp capabilities, you can look at the `"man"` page (shorthand for "manual"). In fact, **every** UNIX command-line program has a built-in `man` page that you can call up to help you. Just type `man` and then the program name and it will give you the manual (small excerpt shown below).

```
[srkeller@pbio381 mydata]$ man grep
```

- One of the most useful aspects of UNIX is the ability to take the output from one command and pass it along as standard input (termed 'stdin') into another command without having to save the intermediate files. This is called "piping", and makes use of the pipe character (|) located above the return key to feed data between programs.

- Example: Say we wanted to create tally of many samples correspond to the Edge region.  We can use `grep` to do the search and use the pipe (|) to send the results of grep to the `wc -l` command, which will tally up the number of lines.

```
[srkeller@pbio381 mydata]$ grep -w "E" RS_Exome_metadata.txt | wc -l
110
```  
  
- Looks like there's 110 sample individuals in the Edge region.  We can get fancier...let's find the # of unique populations that are contained in the edge.  We can do this by piping together multiple commands in a row.  The new commands are: `cut` to get just the column (also known as a "field") containing the population code (the first col, "-f1""), and `uniq` to collapse the number of rows to just 1 per population.  Here's the whole thing in one go:

```
[srkeller@pbio381 mydata]$ grep -w "E" RS_Exome_metadata.txt | cut -f1 | uniq 
AB
BFA
BRB
CR
CRA

[...]
```

- How many Edge pops are there in total?  Just add the `wc -l` command to the end of your piped command:

```
[srkeller@pbio381 mydata]$ grep -w "E" RS_Exome_metadata.txt | cut -f1 | uniq | wc -l
```

- Let's save the list of unique populations to a new file that you name "EdgePops.txt".  How would you do this?

- Now, what if we want to do operations on multiple files at a time?  For example, we've generated three .txt files in our `mydata/` folder, all of which represent metadata that we may want to keep separate from other data files we work with down the road.  There's a way to do this quickly using the wildcard character `*`. With the wildcard, the `*` takes the place of any character, and in fact any length of characters.

  - Let's make a new folder called "metadata" and then move all the text files we've got so far into it:

```
[srkeller@pbio381 mydata]$ mkdir metadata
[srkeller@pbio381 mydata]$ mv *txt metadata/
[srkeller@pbio381 mydata]$ ll metadata/
```

- OK, what about when we have files we don't want anymore? How do we clean up our workspace? You can remove files and folders with the `rm` command. However, in its default mode, UNIX will not ask if you really mean it before getting rid of it forever(!), so this can be dangerous if you're not paying attention. 
  - As an example, let's use our `grep` command to pull out the samples that belong to the "AB"" population and save it to a new file called `metatdata/AB.txt`. But perhaps we later decide we're not going to work with those samples, so we use `rm` to delete that file:

```
[srkeller@pbio381 mydata]$ cd metadata/
[srkeller@pbio381 mydata]$ rm AB.txt 

```

- NOTE: this is equivalent to `rm metadata/AB.txt`
  
- Gone...Forever! If that seems worrisome, you can change your personal settings so that the server asks you to confirm deletion before it acts. To do this, we'll need to follow a couple of new steps:

1. `cd` to your home directory (~/)
2. list all the files, including "hidden" ones that aren't usually shown. To do this, use `ll -a`.
3. Look for a file called ".bashrc" — this contains your settings for how you interact with the server when you log in.
4. We're going to open this file and edit it to add a setting to request that `rm` confirms deletion with us. To edit text files on the fly in UNIX, you can use the built-in text editor, "vim": `vim .bashrc`
5. You should see something that looks like this:

```
  # .bashrc

  # Source global definitions
  if [ -f /etc/bashrc ]; then
          . /etc/bashrc
  fi

  # Uncomment the following line if you don't like systemctl's auto-paging feature:
  # export SYSTEMD_PAGER=

  # User specific aliases and functions
```

6.   Use your arrow key to move your cursor down to the last line, below ""# User specific aliases and functions" — this is where we're going to insert our new function.

7.   By default, vim is in read-only mode when it opens files. To go into edit mode, press your "i" key (for "insert"). You are now able to make changes to the file.

8.   Add the following text on a new line directly below the "# User specific…" line:

       `alias rm='rm -i'`

9.   Your file should now look like this:

```bash
  # .bashrc

  # Source global definitions
  if [ -f /etc/bashrc ]; then
          . /etc/bashrc
  fi

  # Uncomment the following line if you don't like systemctl's auto-paging feature:
  # export SYSTEMD_PAGER=

  # User specific aliases and functions

  alias rm='rm -i'
```

10.    You're now ready to escape out of edit mode (hit the `escape` key), write (save) your changes (type `:w`), and quit vim (type `:q`).  You can also combine this into a single command `:wq`

11.    These changes won't take effect until you log out (type `exit` to log out of the server). But from now on, every time you log in, the server will remember that you want a reminder before deleting any of your work.

-OK, last step:  we need to add all our changes to our repo, commit them, then push them to the server:

```
[srkeller@pbio381 mydata]$ git add --all .
[srkeller@pbio381 mydata]$ git commit -m "comment about your commit"
[srkeller@pbio381 mydata]$ git push
```

- You can check your repo on the Github website to verify that everythings now up to date.

##Let's review what we've learned so far…##

- Logging in to the server: `ssh netid@pbio381.uvm.edu`
- Finding what directory you're in: `pwd`
- Listing files in your current directory, or changing to a new directory: `ll`, `cd`
- Cloning our git repo `git clone`
- Making a new folder: `mkdir foldername`
- Location of shared data on our class server: `/data/project_data/RS_ExomeSeq`
- Copying or moving files from one location to another: `cp filename destinationpath/` or `mv filename destinationpath/` 
- Peeking into the first or last few lines of a file: `head filename`, `tail filename`
- Searching within a file for a match: `grep 'search string' filename`
- Outputing the results of a command to a new file: `grep 'search string' filename >outputfilename`
- Using wildcards to work on multiple files at the same time: `mv *.txt newfolder/`
- Using the "pipe" to send the output of one command to the input of another: `grep 'text' filename | wc -l`
- Removing files or folders: `rm`
- Editing text files on the server: `vim filename`  
- Using git to add, commit, and push our changes to our GitHub account in the cloud

******************

### Handy [UNIX cheat sheet](https://files.fosswire.com/2007/08/fwunixref.pdf) for helping to remember some of these commonly used commands (and others)

### Here's another useful [UNIX cheatsheet](http://cheatsheetworld.com/programming/unix-linux-cheat-sheet/)

### A cheatsheet for [using git on the command line](https://github.github.com/training-kit/downloads/github-git-cheat-sheet.pdf)


