---
title: "P/BIO381 Tutorials: Diversity and Phentyope~Genotype Associations - Day 4"
date: 'February 19, 2020'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

# Learning Objectives for 02/19/20

1. Appreciate the difference between the unfolded and the folded SFS (which should we use?)
2. Calculate diversity stats for our focal pops (SFS, Num sites, Frequency of SNPs, theta-W, pi, Tajima's D)
3. Visualize results in R and share to google drive
4. Introduce Genome-Wide Association Studies (GWAS) in ANGSD using genotype probabilities
5. Do GWAS on seedling heights
 


### The unfolded vs. folded SFS

The big difference here is whether we are confident in the ancestral state of each variable site (SNP) in our dataset

If we know the anestral state, then the best info is contained in the unfolded spectrum, which shows the frequency histogram of how many derived loci are rare vs. common

* bins in the folded spectra go from 0 to 2N -- why?

When you don't know the ancestral state confidently, you can make the SFS based on the minor allele (the less frequent allele; always < 0.5 in the population).  

* bins in the unfolded spectra go from 0 to 1N -- why?

Essentially, we the folded spectra wraps the SFS around such that high frequency "derived" alleles are put in the small bins (low minor allele freq).

![](https://i2.wp.com/evomics.org/wp-content/uploads/2018/01/fig5.png?resize=800%2C571)

### Calculate SFS and diversity stats

In your `myscripts` folder, let's revise the script `ANGSD_mypop.sh` to work on the folded SFS

The first part will be identical as last time, except:
1. Let's change the -out name to `-out ${output}/${mypop}_outFold`
2. Take out the HWE test (not necessary to run again) and replace it with `fold 1`


For info on using the folded spectrum, [see the ANGSD manual page for Theta stats under "unfolded"](http://www.popgen.dk/angsd/index.php/Thetas,Tajima,Neutrality_tests).


#### The above command generated the site allele frequencies, which we can now use to estimate the folded SFS for your pop

```
realSFS ${output}/${mypop}_outFold.saf.idx -maxIter 1000 -tole 1e-6 -P 1 > ${output}/${mypop}_outFold.sfs

```

#### Once you have the SFS, you can estimate the theta diversity stats:

We do this by running ANGSD again, as above, but this time we include the `-pest` flag which uses the SFS we estimated previously as a prior on the allele frequency estimation. We also include the `doThetas` flag, and of course the `fold 1` to tell ANGSD we want the diversities based on the folded SFS

So, we're copying our previous set of ANGSD commands to make a new entry and adding:

```
-pest ${output}/${mypop}_outFold.sfs \
-doSaf 1 \
-doThetas 1 \
-fold 1

thetaStat do_stat ${output}/${mypop}_outFold.thetas.idx

```

The first column of the results file (${mypop}.thetas.idx.pestPG) is formatted a bit funny and we don't really need it. We can use the `cut` command to get rid of it if we want to, or just ignore it.

```
cut -f2- ${mypop}.thetas.idx.pestPG > ${mypop}.thetas
```
This is now ready to bring into R (transfer via git) to look at the mean per-site nucleotide diversity for your focal population.  How does it compare to other populations?  Also bring in your SFS for plotting, and to calculate the SNP frequency in your population.

## Share you population specific stats with the group
Share your population stats on by pasting your values into this [googledoc](https://docs.google.com/spreadsheets/d/1Q8yLVPTki3i3wy8PcFKeYyRUJxp5qNhL29Ytqqky768/edit?usp=sharing).

## Genome-Wide Association Studies (GWAS)

The last type of analysis we'll look at is GWAS.  Recall that the purpose of GWAS is to 'scan' the genome SNP by SNP and test for an association between the genotypes at that locus (e.g., AA, AC, CC) and some response variable of interest.  Often this is a phenotypic trait that has been carefully measured so as to reduce the amount of environmental (non-genetic) contribution.  

* We'll use GWAS to map seedling height growth data in common gardens that my group planted in April-May of 2019 at three sites:  Burlington VT, Frostburg MD, and Bent Creek, NC

*Here are some pics from the VT garden planting...*

![](http://www.uvm.edu/~kellrlab/uploads/2/2/9/6/22961126/editor/ekwdhm1g.jpeg?1564585477)
![](http://www.uvm.edu/~kellrlab/uploads/2/2/9/6/22961126/published/efcbkfma.jpeg?1564592437)
![](http://www.uvm.edu/~kellrlab/uploads/2/2/9/6/22961126/published/xenkbx7o.jpeg?1563386477)

Each family is represented by 5 seedlings per site, randomized within each block

Let's start with a GWAS that maps phenotype~genotype associations in the MD garden, since that is close to the "local" climate environment for many of these edge populations.  I've generated **best linear unbiased predictors (BLUPs)** for each family at the MD site based on an initial ANOVA that removed effects of block and initial height at planting.  The values are here:

`/data/project_data/RS_ExomeSeq/ANGSD/blups/Ht_EDGE_MD.blups`

The ANGSD command we'll use is called `doAsso` and there's some [good guidance on the manual page](http://www.popgen.dk/angsd/index.php/Association) for the different implementations of this amnalysis.  We'll use the hybrid approach `doAsso 5` which tests every SNP for association, but then follows up with only the significant SNPs to estimate the effect size (slope of the phenotype~genotype relationship)

The statistical approach is a generalized linear model of the form:

`y ~ u + beta*SNP + covariates + error`

where, 

* y = the reponse variable (in this case the BLUPs for height, 1 for each family)
* u = the intercept
* beta = the coefficient (or slope) relating the change in the response variable to the SNP genotype
* SNP = an individual SNP locus, as a factor with 3 levels (e.g., AA, AC, CC)
* covariates are 1 or more variables that are needed to help control the type 1 error rate (false positives) Typically one uses PC axes or Admixture ancestry coefficients for this (remember the Norway spruce paper?).  We can use the 1st two PC scores from our PCA, even though the structure seemed very minimal. You can find the scores here: 
```
[srkeller@pbio381 PCA]$ pwd
/data/project_data/RS_ExomeSeq/ANGSD/PCA
[srkeller@pbio381 PCA]$ ll EDGE_PC*
-rw-r--r--. 1 srkeller users 4342 Feb 18 21:01 EDGE_PC12.txt
-rw-r--r--. 1 srkeller users 2193 Feb 18 18:10 EDGE_PC1.txt
-rw-r--r--. 1 srkeller users 2149 Feb 18 18:10 EDGE_PC2.txt
```
* error = the error variance


ANGSD uses genotype probabilities as input and tests each SNP for association with a Likelihood Ratio Test (LRT) which tests the likelihood of there being an association / likelihood of NO association.  The LRT is ~ chisq distributed with 1 d.f.

The basic set up is:

1. Call ANGSD with a list of all the bam files and apply the normal filtering as above to reduce noise (poor mapping or read coverage)
2. Ask ANGSD to estimate genotype probabilities to pass to `doAsso`
3. Tell ANGSD the name of your trait file and covariates file if you are using one.
4. Other options as needed to minimze type 1 error rates -- [see guidance at bottom of ANGSD manual page](http://www.popgen.dk/angsd/index.php/Association)
5. Bring results into R for plotting and outlier detection

#### Let's define our directories first, then layer in the main body of the script

```

blups="/data/project_data/RS_ExomeSeq/ANGSD/blups"

OUT="/data/project_data/RS_ExomeSeq/ANGSD"

```
To speed things up in class a bit, we're going to break the total number of contigs into 20 equal size chunks of 1280.  That way, each student runs their GWAS on a set of 1280 contigs, and then we pool our result together at the end.  No need to do this if you're working on your own in the future...this is just a convenience for class time.

```
contigs="/data/project_data/RS_ExomeSeq/ANGSD/contig_splits"

mycontigs="xaa"

```

And then the main script:

```
ANGSD -b ${OUT}/EDGE_bam.list \
-ref ${REF} 
-out ${OUT}/GWAS/Ht_EDGE_MD_PC12_${mycontigs} \
-remove_bads 1 \
-C 50 \
-baq 1 \
-minMapQ 20 \
-minQ 20 \
-setMinDepth 3 \
-minInd 2 \
-setMinDepthInd 1 \
-setMaxDepthInd 17 \
-skipTriallelic 1 \
-GL 1 \
-doPost 1 \
-doCounts 1 \
-doMajorMinor 1 \
-doMaf 1 \
-SNP_pval 1e-6 \
-yQuant ${blups}/Ht_EDGE_MD.blups \
-doAsso 5 \
-rf ${contigs}/${mycontigs} \
-cov ${OUT}/PCA/EDGE_PC12.txt

```

Couple of new options here:

* -yQuant ${blups}/Ht_EDGE_MD.blups  -- this is your response variable; no header
* -doAsso 5 -- this is the model you're choosing to run.  See ANGSD manual under doAsso
* -rf ${contigs}/${mycontigs} -- this means you want the GWAS to run just on a set of regions (r) defined in an external file (f)
* -cov ${OUT}/PCA/EDGE_PC12.txt -- these are the covariates; as many columns as you like; no header


Give it some time to run (should take 20 min or so), then we'll bring the results into R to plot and summarize.

* Is there evidence for genetic associations with early seedling height growth?
* What do the beta estimates suggest about the direction of the effect (more often increasing, decreasing)?
* Are SNPs that have a positive effect on height in the MD garden the same as those in NC?
* For the SNPs that have the strongest effect on height, are they polymorphic in your focal pop?  

#### NOTE:  You can do the same analysis replacing climate data for heigh as the response variable.  This is known as Gene-Environment Association (GEA), and is a common way to test for regions of the genome that confer local adaptation to climate (or other aspects of the environment).  I've placed two relevant climate vareiables for red spruce in the blups directory:

```
[srkeller@pbio381 blups]$ pwd
/data/project_data/RS_ExomeSeq/ANGSD/blups
[srkeller@pbio381 blups]$ ll
total 20
-rw-r--r--. 1 srkeller users  432 Feb 18 20:41 bio18.blups
-rw-r--r--. 1 srkeller users  432 Feb 18 20:37 bio5.blups
```

These variables are derived from the [bioclim climatic variables available here](https://www.worldclim.org/bioclim).

* Bio5 = 10*(Max temp of the warmest month)
* Bio18 = 10*(Precip of the warmest quarter)




